oqamaint/openqabot.py

```diff
@@ -0,0 +1,588 @@
+
+        buildnr = 0
+
+        # not found, then check for the next free build nr
+        for job in jobs:
+            build = job['settings']['BUILD']
+            if build and build.startswith(today):
+                try:
+                    nr = int(build.split('-')[1])
+                    if nr > buildnr:
+                        buildnr = nr
+                except ValueError:
+                    continue
+
+        buildnr = "{!s}-{:d}".format(today, buildnr + 1)
+
+        for s in data['settings']:
+            # now schedule it for real
+            if 'incidents' in data.keys():
+                for x, y in self.calculate_incidents(data['incidents']):
+                    s[x] = y
+            s['BUILD'] = buildnr
+            s['REPOHASH'] = repohash
+            self.logger.debug("Prepared: {}".format(pformat(s)))
+            if not self.dryrun:
+                try:
+                    self.logger.info("Openqa isos POST {}".format(pformat(s)))
+                    self.openqa.openqa_request('POST', 'isos', data=s, retries=1)
+                except Exception as e:
+                    self.logger.error(e)
+        self.update_test_builds[prj] = buildnr
+
+    def request_get_openqa_jobs(self, req, incident=True, test_repo=False):
+        ret = None
+        types = {a.type for a in req.actions}
+        if 'maintenance_release' in types:
+            src_prjs = set([a.src_project for a in req.actions])
+            if len(src_prjs) != 1:
+                raise Exception("can't handle maintenance_release from different incidents")
+            build = src_prjs.pop()
+            tgt_prjs = set([a.tgt_project for a in req.actions])
+            ret = []
+            if incident:
+                ret += self.openqa_jobs[build]
+            for prj in sorted(tgt_prjs):
+                repo_settings = self.tgt_repo.get(self.openqa.baseurl, {})
+                if test_repo and prj in repo_settings:
+                    repo_jobs = self.openqa_jobs[prj]
+                    ret += repo_jobs
+
+        return ret
+
+    def calculate_qa_status(self, jobs=None):
+        if not jobs:
+            return QA_UNKNOWN
+
+        j = {}
+        has_failed = False
+        in_progress = False
+
+        for job in jobs:
+            if job['clone_id']:
+                continue
+            name = job['name']
+
+            if name in j and int(job['id']) < int(j[name]['id']):
+                continue
+            j[name] = job
+
+            if job['state'] not in ('cancelled', 'done'):
+                in_progress = True
+            else:
+                if job['result'] != 'passed' and job['result'] != 'softfailed':
+                    has_failed = True
+
+        if not j:
+            return QA_UNKNOWN
+        if in_progress:
+            return QA_INPROGRESS
+        if has_failed:
+            return QA_FAILED
+
+        return QA_PASSED
+
+    def add_comment(self, msg, state, request_id=None, result=None):
```

Comment:
See `ReviewBot.write_comment()` and related note on comment regex at the top of this file. Should be able to drop all of this.

Reply:
"""Write comment from log messages if not similar to previous comment.""" ... comments from logger ? really ? No way

Reply:
In the time you spent to write this comment you could have read the method and seen that it provides everything you need. I am not sure why you are so extremely resistant. This code-base is far from ideal, but I've refactored a lot to make this possible. The comment does not cover all the aspects of the function.

`leaper.py` was built that way so this function started with that as base functionality.

Reply:
Added some long-form documentation: https://github.com/openSUSE/osc-plugin-factory/pull/1154/commits/76c4b499e317aa7f24d199e9cbf3871b4d42c12f

Reply:
still "logging" purpose isn't to hold state ...

Reply:
a) indeed, it was shortcut make in `leaper`, not by me.
b) unrelated to the topic at hand since you can pass `message` parameter.